### PR TITLE
feat: for-direction rule add check for condition in reverse order

### PR DESCRIPTION
**Pull request by ild0tt0re**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

**What rule do you want to change?**
- [for-direction rule](https://eslint.org/docs/latest/rules/for-direction)

**What change do you want to make (place an "X" next to just one item)?**

- [x] Generate more warnings
- [ ] Generate fewer warnings
- [ ] Implement autofix
- [ ] Implement suggestions

**How will the change be implemented (place an "X" next to just one item)?**

- [ ] A new option
- [x] A new default behaviour
- [ ] Other

**Please provide some example code that this change will affect:**

```js
for (var i = 0; 10 > i; i--) {
}

for (var i = 0; 10 >= i; i--) {
}

for (var i = 10; 10 < i; i++) {
}

for (var i = 10; 0 <= i; i++) {
}
```

**What does the rule currently do for this code?**

The `for-direction` rule checks **only** for the condition with the counter i.e. `i` on the **_left side_** of the binary expression `i < 10` like:
```js
for (var i = 0; i < 10; i++) {
}
``` 

**What will the rule do after it's changed?**

The rule adds more warnings for code like this with the counter i.e. `i` on the **_right side_** of the test condition `10 > i`
```js
for (var i = 0; 10 > i; i++) {
}
```

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Hi @ild0tt0re, thanks for the PR!

Could you please copy & paste [this template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md) into https://github.com/eslint/eslint/pull/17755#issue-1990992833 and fill it out? We'll need those details to understand and evaluate the proposal.

**Comment by ild0tt0re** [comment]
> Hi @ild0tt0re, thanks for the PR!
> 
> Could you please copy & paste [this template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md) into [#17755 (comment)](https://github.com/eslint/eslint/pull/17755#issue-1990992833) and fill it out? We'll need those details to understand and evaluate the proposal.

Thanks to you!
Done ✅

**Comment by mdjermanovic** [comment]
So, the `for-direction` rule currently reports `for (var i = 0; i < 10; i--) {}` but doesn't report `for (var i = 0; 10 > i; i--) {}` ([Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGZvci1kaXJlY3Rpb246IFwiZXJyb3JcIiAqL1xuXG5mb3IgKHZhciBpID0gMDsgaSA8IDEwOyBpLS0pIHt9XG5cbmZvciAodmFyIGkgPSAwOyAxMCA+IGk7IGktLSkge30iLCJvcHRpb25zIjp7ImVudiI6eyJlczYiOnRydWV9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==)), although they're basically the same.

I support the proposal to account for conditions in this style, especially since we have a rule ([`yoda`](https://eslint.org/docs/latest/rules/yoda)) that can be configured to enforce it. For example, in [this demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGZvci1kaXJlY3Rpb246IFwiZXJyb3JcIiAqL1xuLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwiYWx3YXlzXCJdICovXG5cbmZvciAodmFyIGkgPSAwOyBpIDwgMTA7IGktLSkge31cbiIsIm9wdGlvbnMiOnsiZW52Ijp7ImVzNiI6dHJ1ZX0sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQifX19), `for-direction` reports an error. But after the `yoda` fix is applied, the `for-direction` error disappears although the same bug in the code remains.

Just waiting for more opinions from the team before marking this as accepted and reviewing it.

**Comment by fasttime** [comment]
> I support the proposal to account for conditions in this style, especially since we have a rule ([`yoda`](https://eslint.org/docs/latest/rules/yoda)) that can be configured to enforce it. For example, in [this demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGZvci1kaXJlY3Rpb246IFwiZXJyb3JcIiAqL1xuLyogZXNsaW50IHlvZGE6IFtcImVycm9yXCIsIFwiYWx3YXlzXCJdICovXG5cbmZvciAodmFyIGkgPSAwOyBpIDwgMTA7IGktLSkge31cbiIsIm9wdGlvbnMiOnsiZW52Ijp7ImVzNiI6dHJ1ZX0sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQifX19), `for-direction` reports an error. But after the `yoda` fix is applied, the `for-direction` error disappears although the same bug in the code remains.

I agree.

File: lib/rules/for-direction.js [hunk 1]
```diff
@@ -101,20 +101,36 @@ module.exports = {
...
+         * get the counter position in test
+         * @param {ASTNode} node the node to check
+         * @returns {"left" | "right"} the position of the counter respect to BinaryExpression
+         */
+        function getCounterPosition(node) {
+            if (node.update.type === "UpdateExpression" || node.update.type === "AssignmentExpression") {
+                if (node.update.argument && node.update.argument.type === "Identifier") {
+                    return node.test.left.name === node.update.argument.name ? "left" : "right";
+                }
+            }
```
Comment (mdjermanovic) [inline lib/rules/for-direction.js:115]: This doesn't seem to cover AssignmentExpression:

```js
for(var i = 0; 10 > i; i-=1){} // should be error
```

and looks partially redundant with functions that check whether the names match afterward.

What do you think about updating ForStatement(node) to always check both sides instead:

```js
ForStatement(node) {

    if (node.test && node.test.type === "BinaryExpression" && node.update) {
        for (const counterPosition of ["left", "right"]) {
            if (node.test[counterPosition].type !== "Identifier") {
                continue;
            }

            const counter = node.test[counterPosition].name;
            // ...
        }
    }
}
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/for-direction.js (hunk 1 above)
Comment (ild0tt0re) [reply lib/rules/for-direction.js]: Yes much better! Do you think it is necessary to add other tests on top of this? I added only one more 
`{ code: "for(var i = 0; 10 > i; i-=1){}", errors: [incorrectDirection] }` after your suggestion.

Comment (mdjermanovic) [reply lib/rules/for-direction.js]: > Do you think it is necessary to add other tests on top of this? I added only one more
> `{ code: "for(var i = 0; 10 > i; i-=1){}", errors: [incorrectDirection] }` after your suggestion.

One test looks fine to me.

**Pull request merged**